### PR TITLE
Fixed onError event dispatching

### LINKED ISSUE #71: Form::fireEvents() calls onError only when it has multiple success handlers
it should be something like this

```php
if ($this->onSuccess) {
	foreach ($this->onSuccess as $handler) {
		if (!$this->isValid()) {
			break;
		}
		$params = Nette\Utils\Callback::toReflection($handler)->getParameters();
		$values = isset($params[1]) ? $this->getValues($params[1]->isArray()) : NULL;
		Nette\Utils\Callback::invoke($handler, $this, $values);
	}
}

if (!$this->isValid()) {
	$this->onError($this);
}
```

because now, when I have only one success handler and add errors in it, the onError is not called

### PR DESCRIPTION
**Pull request by enumag**
This closes #71.

### CODE REVIEW

File: tests/Forms/Forms.onSuccess.phpt [hunk 1]
```diff
@@ -34,8 +33,23 @@ $form->onSuccess[] = function() use (& $called) {
...
 };
+$form->fireEvents();
+Assert::same(array(1, 2, 'err'), $called);
 
 
+$called = array();
+$form = new Form;
+$form->addText('name');
+$form->addSubmit('submit');
+$form->onSuccess[] = function() use (& $called) {
```
Comment (fprochazka) [inline tests/Forms/Forms.onSuccess.phpt]: tady ten by tu být neměl aby se ověřilo že se to zavolá ikdyž je tam jenom jeden onSuccess

Comment (enumag) [inline tests/Forms/Forms.onSuccess.phpt]: nn, ono se to nevolalo pokud error nastal v posledním onSuccess handleru - jejich celkový počet nemá vliv

Comment (fprochazka) [inline tests/Forms/Forms.onSuccess.phpt]: Pravda :)

**Comment by enumag** [comment]
Hmm... the onError event won't be called if I call the addError on a field and not the form itself.

**Comment by enumag** [comment]
Should I refactor it to this? It has better performance because isValid is not called multiple times.

```php
		if (!$this->isValid()) {
			$this->onError($this);
		} elseif ($this->onSuccess) {
			foreach ($this->onSuccess as $handler) {
				$params = Nette\Utils\Callback::toReflection($handler)->getParameters();
				$values = isset($params[1]) ? $this->getValues($params[1]->isArray()) : NULL;
				Nette\Utils\Callback::invoke($handler, $this, $values);
				if (!$this->isValid()) {
					$this->onError($this);
					break;
				}
			}
		}
```

**Comment by dg** [comment]
Can you write better commit message?

**Comment by enumag** [comment]
What's wrong with this one? I can't come up with anything better.

**Comment by vojtech-dobes** [comment]
`Form::fireEvents() calls onError even with only one success handler`

**Comment by enumag** [comment]
@vojtech-dobes That's actually wrong. The onError event previously wasn't called if the error was added by last onSuccess handler (regardless how many there are in total).

**Comment by enumag** [comment]
Maybe `Form::fireEvents() can call onError even after last onSuccess handler`

**Comment by fprochazka** [comment]
@enumag :+1:

**Comment by enumag** [comment]
Commit message changed (thx @fprochazka). I've also changed the implementation to the one with better performance as mentioned above.

**Comment by vojtech-dobes** [comment]
:+1:

**Comment by dg** [comment]
Thank you

**Pull request merged**